irohad/ordering/impl/ordering_gate_impl.cpp

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ordering/impl/ordering_gate_impl.hpp"
+#include <grpc++/grpc++.h>
+
+namespace iroha {
+  namespace ordering {
+
+    OrderingGateImpl::OrderingGateImpl(const std::string &server_address)
+        : thread_(&OrderingGateImpl::asyncCompleteRpc, this) {
+      stub_ = proto::OrderingService::NewStub(grpc::CreateChannel(
+          server_address, grpc::InsecureChannelCredentials()));
+    }
+
+    void OrderingGateImpl::propagate_transaction(
+        const model::Transaction &transaction) {
+      auto call = new AsyncClientCall;
+
+      call->response_reader = stub_->AsyncSendTransaction(
+          &call->context, factory_.serialize(transaction), &cq_);
+
+      call->response_reader->Finish(&call->reply, &call->status, call);
+    }
+
+    rxcpp::observable<model::Proposal> OrderingGateImpl::on_proposal() {
+      return proposals_.get_observable();
+    }
+
+    grpc::Status OrderingGateImpl::SendProposal(
+        ::grpc::ServerContext *context, const proto::Proposal *request,
+        ::google::protobuf::Empty *response) {
+      // auto removes const qualifier of model::Proposal.transactions
+      auto transactions =
+          decltype(std::declval<model::Proposal>().transactions)();
+      for (const auto &tx : request->transactions()) {
+        transactions.push_back(factory_.deserialize(tx));
+      }
+
+      handleProposal(model::Proposal(transactions));
+
+      return grpc::Status::OK;
+    }
+
+    void OrderingGateImpl::handleProposal(model::Proposal &&proposal) {
+      proposals_.get_subscriber().on_next(proposal);
+    }
+
+    void OrderingGateImpl::asyncCompleteRpc() {
+      void *got_tag;
+      auto ok = false;
+      while (cq_.Next(&got_tag, &ok)) {
+        auto call = static_cast<AsyncClientCall *>(got_tag);
```

Comment:
Do you mean accessing `reply` or `status` of `AsyncClientCall`? We don't care about the reply here, since we simulate publishing here.